[GTK] Honor GtkSettings:gtk-font-rendering setting

carlosgcampos opened the pull request:
#### 3172ac2bd7ab8d25a76213d488e515ef0b45baf7
<pre>
[GTK] Honor GtkSettings:gtk-font-rendering setting
<a href="https://bugs.webkit.org/show_bug.cgi?id=281665">https://bugs.webkit.org/show_bug.cgi?id=281665</a>

Reviewed by Adrian Perez de Castro.

And ignore the font system settings when rendering mode is set to automatic.

* Source/WebCore/platform/glib/SystemSettings.cpp:
(WebCore::SystemSettings::updateSettings):
* Source/WebCore/platform/glib/SystemSettings.h:
(WebCore::SystemSettings::followFontSystemSettings const):
* Source/WebCore/platform/graphics/FontRenderOptions.h:
(WebCore::FontRenderOptions::setFollowSystemSettings):
(WebCore::FontRenderOptions::hinting const): Deleted.
(WebCore::FontRenderOptions::antialias const): Deleted.
(WebCore::FontRenderOptions::subpixelOrder const): Deleted.
* Source/WebCore/platform/graphics/skia/FontRenderOptionsSkia.cpp:
(WebCore::FontRenderOptions::hinting const):
(WebCore::FontRenderOptions::antialias const):
(WebCore::FontRenderOptions::subpixelOrder const):
* Source/WebKit/Shared/glib/SystemSettings.serialization.in:
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.cpp:
(WebKit::SystemSettingsManagerProxy::settingsDidChange):
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.h:
* Source/WebKit/UIProcess/gtk/SystemSettingsManagerProxyGtk.cpp:
(WebKit::SystemSettingsManagerProxy::followFontSystemSettings const):
(WebKit::SystemSettingsManagerProxy::SystemSettingsManagerProxy):
* Source/WebKit/WebProcess/glib/SystemSettingsManager.cpp:
(WebKit::SystemSettingsManager::SystemSettingsManager):

Canonical link: <a href="https://commits.webkit.org/285502@main">https://commits.webkit.org/285502@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19bbf55a6b352ef7558d946ca07e1919481b8b8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25671 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77069 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74986 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23921 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57300 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15786 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75938 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62722 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37725 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43926 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20188 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22434 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65781 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78742 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17117 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19689 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65744 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17165 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62728 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65018 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13333 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6990 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11197 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48094 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2881 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49161 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50456 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48906 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/FontRenderOptions.h:
```diff
@@ -39,6 +39,12 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
 WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
 #endif
 
+#if PLATFORM(GTK) && !USE(GTK4)
+static constexpr bool s_followSystemSettingsDefault = true;
+#else
+static constexpr bool s_followSystemSettingsDefault = false;
+#endif
```
What's the explanation for different behavior in GTK 3 vs. GTK 4?

carlosgcampos replied:
I don't know, I'm just being consistent with GTK here.

mxschmitt replied:
This caused a rendering difference for Playwright between WPE and GTK when wrapping text. In `FontRenderOptions::hinting` it was before consistent but now on GTK its `m_followSystemSettings=true` so it uses the system setting and on WPE it falls back to `SkFontHinting::kSlight` which is different to `SkFontHinting::kNormal`. Is this worth filing an issue?

Test: https://github.com/microsoft/playwright/blob/487134fbafcd8840fd1caf34443578608f0cc92d/tests/library/headful.spec.ts#L287-L315

Edit: Seems like the easiest way of aligning it for us is to return `false` in SystemSettingsManagerProxy::followFontSystemSettings.

mcatanzaro replied:
I guess my question is: why would you expect WebKitGTK and WPE to have consistent font rendering? WPE notably cannot look at your GTK/cairo settings, which have a large influence on the result.

carlosgcampos committed:
[GTK] Honor GtkSettings:gtk-font-rendering setting
https://bugs.webkit.org/show_bug.cgi?id=281665

Reviewed by Adrian Perez de Castro.

And ignore the font system settings when rendering mode is set to automatic.

* Source/WebCore/platform/glib/SystemSettings.cpp:
(WebCore::SystemSettings::updateSettings):
* Source/WebCore/platform/glib/SystemSettings.h:
(WebCore::SystemSettings::followFontSystemSettings const):
* Source/WebCore/platform/graphics/FontRenderOptions.h:
(WebCore::FontRenderOptions::setFollowSystemSettings):
(WebCore::FontRenderOptions::hinting const): Deleted.
(WebCore::FontRenderOptions::antialias const): Deleted.
(WebCore::FontRenderOptions::subpixelOrder const): Deleted.
* Source/WebCore/platform/graphics/skia/FontRenderOptionsSkia.cpp:
(WebCore::FontRenderOptions::hinting const):
(WebCore::FontRenderOptions::antialias const):
(WebCore::FontRenderOptions::subpixelOrder const):
* Source/WebKit/Shared/glib/SystemSettings.serialization.in:
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.cpp:
(WebKit::SystemSettingsManagerProxy::settingsDidChange):
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.h:
* Source/WebKit/UIProcess/gtk/SystemSettingsManagerProxyGtk.cpp:
(WebKit::SystemSettingsManagerProxy::followFontSystemSettings const):
(WebKit::SystemSettingsManagerProxy::SystemSettingsManagerProxy):
* Source/WebKit/WebProcess/glib/SystemSettingsManager.cpp:
(WebKit::SystemSettingsManager::SystemSettingsManager):

Canonical link: https://commits.webkit.org/285502@main
